File: myulib-core/src/main/java/com/myudog/myulib/api/core/animation/Easing.java
package com.myudog.myulib.api.core.animation;

public enum Easing {
    LINEAR {
        @Override
        public double apply(double progress) {
            return clamp(progress);
        }
    },
    EASE_IN_QUAD {
        @Override
        public double apply(double progress) {
            double t = clamp(progress);
            return t * t;
        }
    },
    EASE_OUT_QUAD {
        @Override
        public double apply(double progress) {
            double t = clamp(progress);
            return t * (2.0 - t);
        }
    },
    EASE_IN_OUT_QUAD {
        @Override
        public double apply(double progress) {
            double t = clamp(progress);
            if (t < 0.5) {
                return 2.0 * t * t;
            }
            return 1.0 - Math.pow(-2.0 * t + 2.0, 2.0) / 2.0;
        }
    },
    EASE_IN_CUBIC {
        @Override
        public double apply(double progress) {
            double t = clamp(progress);
            return t * t * t;
        }
    },
    EASE_OUT_CUBIC {
        @Override
        public double apply(double progress) {
            double t = clamp(progress) - 1.0;
            return t * t * t + 1.0;
        }
    },
    EASE_IN_OUT_CUBIC {
        @Override
        public double apply(double progress) {
            double t = clamp(progress);
            if (t < 0.5) {
                return 4.0 * t * t * t;
            }
            return 1.0 - Math.pow(-2.0 * t + 2.0, 3.0) / 2.0;
        }
    },
    SMOOTH_STEP {
        @Override
        public double apply(double progress) {
            double t = clamp(progress);
            return t * t * (3.0 - 2.0 * t);
        }
    };

    public abstract double apply(double progress);

    public float apply(float progress) {
        return (float) apply((double) progress);
    }

    public static double clamp(double progress) {
        if (progress < 0.0) {
            return 0.0;
        }
        if (progress > 1.0) {
            return 1.0;
        }
        return progress;
    }
}

File: src/main/kotlin/com/myudog/myulib/api/animation/Easing.kt
package com.myudog.myulib.api.animation

import kotlin.math.pow
import kotlin.math.sin
import kotlin.math.PI

enum class Easing {
    LINEAR {
        override fun calculate(t: Double) = t
    },
    QUAD_OUT {
        override fun calculate(t: Double) = 1 - (1 - t) * (1 - t)
    },
    ELASTIC_OUT {
        override fun calculate(t: Double): Double {
            val c4 = (2 * PI) / 3
            return if (t == 0.0) 0.0 else if (t == 1.0) 1.0
            else 2.0.pow(-10 * t) * sin((t * 10 - 0.75) * c4) + 1
        }
    };

    abstract fun calculate(t: Double): Double
}